File: zenoh-grpc-client-sdk/zenoh-grpc-cpp/examples/queryable.cpp
#include <iostream>

#include "common.hpp"

int main(int argc, char** argv) {
    auto session = zenoh_grpc::Session::connect(example_endpoint(argc, argv));

    zenoh_grpc::Session::QueryableOptions options;
    options.complete = false;
    options.allowed_origin = zenoh_grpc::Locality::Any;
    auto queryable = session.declare_queryable("demo/query/**", options);

    std::cout << "queryable ready on " << example_endpoint(argc, argv) << std::endl;
    while (true) {
        auto query = queryable.recv();
        std::cout << "query id=" << query.query_id() << " selector=" << query.selector()
                  << " key=" << query.key_expr() << " parameters=" << query.parameters()
                  << " payload=" << bytes_to_string(query.payload())
                  << " encoding=" << query.encoding() << std::endl;

        zenoh_grpc::Query::ReplyOptions reply_options;
        reply_options.encoding = "text/plain";

        const std::string reply1 = "this is a reply1 from c++ queryable";
        query.reply(query.key_expr(),
                    std::vector<std::uint8_t>(reply1.begin(), reply1.end()),
                    reply_options);
        std::cout << "reply 1 sent" << std::endl;

        example_sleep_ms(1000);

        const std::string reply2 = "this is a reply2 from c++ queryable";
        query.reply(query.key_expr(),
                    std::vector<std::uint8_t>(reply2.begin(), reply2.end()),
                    reply_options);
        std::cout << "reply 2 sent" << std::endl;
        
        // The query will be automatically finished when the Query object is destructed, but we can also finish it explicitly here.
        query.finish();
    }

    return 0;
}

File: zenoh-grpc-client-sdk/zenoh-grpc-python/examples/queryable.py
import zenoh_grpc
import time

with zenoh_grpc.Session.connect() as session:
    queryable = session.declare_queryable("demo/query/**", 
                                          callback=None, # if None, you can use recv() to receive queries in a loop, but it is blocking and not recommended. It is better to use a callback to receive queries, as shown in queryable_callback.py.
                                          complete=False,  # optional, default is False. 
                                          allowed_origin=zenoh_grpc.Locality.ANY) # optional, default is ANY.

    while True:
        # not recommended to use recv() in a loop, as it is blocking and will block the main thread, and due to the inner rust implementation, it can not be killed by Ctrl+C. It is better to use a callback to receive samples, as shown in queryable_callback.py. But here we just want to show how to use recv() in a loop, so we use it here for simplicity.
        query = queryable.recv() # blocking, returns a Query object when a query is received
        print("\nreceived query:", query.query_id, query.selector, query.key_expr, query.parameters, query.payload, query.encoding, query.attachment)

        # You can reply multiple times to the same query by calling reply() multiple times. The querier will receive the replies one by one or all at once depending on the consolidation mode set by the querier (AUTO(default,LATEST), NONE(receive one by one), MONOTONIC(time monotonically consolidation), LATEST(only the latest)).

        query.reply(query.key_expr, b"this is a reply1 from queryable", encoding="text/plain")
        print("reply 1 to query:", query.query_id)

        time.sleep(1) # simulate some processing time

        query.reply(query.key_expr, b"this is a reply2 from queryable", encoding="text/plain")
        print("reply 2 to query:", query.query_id)

        # Notice: you should explicitly drop() after you finished replying to send ResponseFinal message to querier, otherwise the querier will wait until timeout and will not receive the replies until then.
        query.drop()
        print("dropped query")
